### PR TITLE
New providers with support AutoExtractHtml

### PR DESCRIPTION
**Pull request by ivanprado**
Many new features in this PR:
* Support for the new providers interface introduced in https://github.com/scrapinghub/scrapy-poet/pull/41
* `AutoExtractHtml` page input support
* `AutoExtractProductListData` support
* Fast stop on cancellation by the use of the `TaskManager`
* Per domain concurrency compliance
* Improved stats reporting

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -239,7 +241,8 @@ When using the AutoExtract providers, be aware that:
 
 * With scrapy-poet integration, retry requests don't go through Scrapy
 * Not all data types are supported with scrapy-poet,
-  currently only Articles and Products are supported
+  currently only Articles Products and Product List are supported with
```
Comment (kmike) [inline README.rst:244]:
```suggestion
  currently only Articles, Products and Product List are supported with
```

File: scrapy_autoextract/errors.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+def summarize_exception(exc):
+    """
+    Provides a text that represents the exception. To be used in stats, so
+    produced text shouldn't be too diverse.
+    """
+    if isinstance(exc, QueryError):
+        msg = exc.message
+        if DomainOccupied.from_message(msg):
+            # Removes variability from domain occupy messages
+            msg = "domain occupied"
```
Comment (kmike) [inline scrapy_autoextract/errors.py:23]: I haven't tried it (maybe it is too noisy), but won't it be interesting to know which domains are causing domain occupied errors?

File: scrapy_autoextract/errors.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+        msg = f"/query/{msg}"
+    else:
+        msg = f"/rest/{exc.__class__.__name__}"
```
Comment (kmike) [inline scrapy_autoextract/errors.py:26]: "rest" => "request"?

File: scrapy_autoextract/providers.py [hunk 4]
```diff
@@ -1,114 +1,216 @@
...
+        self.logger.info(
+            f"AutoExtractProvider started. Retries: {retries_count}, "
+            f"per domain concurrency: {per_domain_concurrency}"
+        )
 
-class ArticleDataProvider(_Provider):
-
-    provided_class = AutoExtractArticleData
+    def create_aiohttp_session(self) -> aiohttp.ClientSession:
+        concurrent_connections = self.settings.get("CONCURRENT_REQUESTS", 16)
```
Comment (kmike) [inline scrapy_autoextract/providers.py:98]: getint?

File: scrapy_autoextract/providers.py [hunk 5]
```diff
@@ -1,114 +1,216 @@
...
+        Return the key that will be used to identify the domain of this request.
+        This key is used to modulate the per request concurrency that can be
+        set using the setting `CONCURRENT_REQUESTS_PER_DOMAIN`.
 
-def install():
-    """Register all providers for their respective provided classes."""
-    ArticleDataProvider.register()
-    ProductDataProvider.register()
+        By default the key is the request domain. Override it to change
+        the behavior.
```
Comment (kmike) [inline scrapy_autoextract/providers.py:124]: I wonder if it should use `request.meta['download_slot']` if it is set

File: scrapy_autoextract/errors.py (hunk 3 above)
Comment (ivanprado) [reply scrapy_autoextract/errors.py:26]: Is `rest` because is the rest of errors. It will be something like:
```
/pages/error/query: 29
/pages/error/rest: 12
```

I'm thinking that I'm going to rename `error` by `errors`

Comment (ivanprado) [reply scrapy_autoextract/errors.py:26]: Done here: https://github.com/scrapinghub/scrapy-autoextract/pull/16/commits/41d3426f1a02ca4f28106dac5e7c5c6d43794dee

File: scrapy_autoextract/providers.py (hunk 4 above)
Comment (ivanprado) [reply scrapy_autoextract/providers.py]: Fixed here: https://github.com/scrapinghub/scrapy-autoextract/pull/16/commits/1059ed4444767b561768df2a80f789ed8894d694

File: scrapy_autoextract/providers.py (hunk 5 above)
Comment (ivanprado) [reply scrapy_autoextract/providers.py:124]: Good point. Let's think about it. This is the Scrapy code that computes the final `download_slot` key:

https://github.com/scrapy/scrapy/blob/0e579182319504ba7bfd0c09333fe92f70c6d312/scrapy/core/downloader/__init__.py#L108-L116

Ideally, we should just mimic the behaviour in the provider, even calling the same function to compute the key. The later doesn't look like the right thing: it would involve accessing the downloader from a provider and invoke a private function. 

We might copy the code directly here in the function as:
```py
def get_per_domain_concurrency_slot(self, request: ScrapyRequest) -> Hashable:
        if 'download_slot' in request.meta:
            return request.meta['download_slot']

        key = urlparse_cached(request).hostname or ''
        if self.ip_concurrency:
            key = dnscache.get(key, key)
        return key
```

`self.ip_concurrency` would be `self.settings.getint('CONCURRENT_REQUESTS_PER_IP')`

If at some point, you need to decouple from this policy, you can just override the provider and implement this method differently. This way `tldextract` wouldn't be a requirement for this project anymore. 

It seems to work. What do you think? I'm going to implement it.

Comment (ivanprado) [reply scrapy_autoextract/providers.py:124]: @kmike I have implemented the logic in a different PR (https://github.com/scrapinghub/scrapy-autoextract/pull/17), because now I'm unsure. Particularly with the per IP logic. Look to this line: https://github.com/scrapinghub/scrapy-autoextract/pull/17/files#diff-64092c6e77be2d563b1b97477967d8306ea8901a9aa46db4023465ecd64d4559R147

This is accessing just a DNS cache... But logic to resolve DNS and populate the cache runs separately (https://docs.scrapy.org/en/latest/topics/settings.html?highlight=dns#std-setting-DNS_RESOLVER). I'm unsure about if we would get AutoExtract URLs resolved and populated automatically there. Do you know it? I would say it is not the case, as we are not using Twisted there for anything. 

So it seems that implementing per IP policy wouldn't be as easy.

Comment (ivanprado) [reply scrapy_autoextract/providers.py:124]: @kmike I'm thinking again on all this. It makes no sense to implement IP policy or `download_slot` here at all. The idea here is to be aligned with what AutoExtract API is using internally for domain occupy, and this is the domain as is defined in this PR. I don't see why it could be useful to implement a different policy here. So +1 to leave the PR as is.

File: scrapy_autoextract/providers.py [hunk 6]
```diff
@@ -1,114 +1,226 @@
...
 
-        request = AutoExtractRequest(
-            url=self.request.url,
-            pageType=page_type,
-        )
-        api_key = self.settings.get("AUTOEXTRACT_USER")
-        endpoint = self.settings.get("AUTOEXTRACT_URL")
-        max_query_error_retries = self.settings.getint(
-            "AUTOEXTRACT_MAX_QUERY_ERROR_RETRIES", 3
+    logger = crawler.spider.logger
```
[the remark this answers is not in the document]
Comment (ivanprado) [reply scrapy_autoextract/providers.py]: I always had a doubt about what was the right thing to do. Now I understand after your explanation. Changed in https://github.com/scrapinghub/scrapy-autoextract/pull/16/commits/d9555870d8898221fb3d296e9a563cba71ff6366

File: scrapy_autoextract/providers.py [hunk 7]
```diff
@@ -1,114 +1,226 @@
...
+    if exception.error_data().get("type") == ACCOUNT_DISABLED_ERROR_TYPE:
+        logger.info(
+            "'Account disabled' request error received. Shutting down the spider"
         )
+        crawler.engine.close_spider(crawler.spider, "account_disabled")
```
[the remark this answers is not in the document]
Comment (ivanprado) [reply scrapy_autoextract/providers.py:54]: Maybe. But I don't want to change that as it could be affecting other services that could be relying on this status.

File: scrapy_autoextract/providers.py [hunk 8]
```diff
@@ -1,114 +1,226 @@
...
-        return cls.provided_class.item_key
+    def provided_classes(cls, type_: Callable) -> bool:
+        return (inspect.isclass(type_)
+                and issubclass(type_, (AutoExtractData, AutoExtractHtml)))
 
+    def __init__(self, crawler: Crawler):
+        """Initialize provider storing its dependencies as attributes."""
+        self.crawler = crawler
+        self.settings = crawler.spider.settings
+        self.logger = self.crawler.spider.logger
```
[the remark this answers is not in the document]
Comment (ivanprado) [reply scrapy_autoextract/providers.py]: Changed

File: scrapy_autoextract/providers.py [hunk 9]
```diff
@@ -1,114 +1,226 @@
...
-    provided_class = AutoExtractArticleData
+    async def close_aiohttp_session(self):
+        if self.aiohttp_session:
+            await self.aiohttp_session.close()
 
+    def create_retry_wrapper(self):
+        return RetryFactory().build()
 
-class ProductDataProvider(_Provider):
+    async def do_request(self, *args, **kwargs):
```
[the remark this answers is not in the document]
Comment (ivanprado) [reply scrapy_autoextract/providers.py:116]: Not sure getting it. `async` and `await` are required as `request_raw` is a coroutine. `args` and `kwargs` are also required as they are carrying arguments (they are not empty). But maybe you are meaning a different thing?

I think that maybe you are referring that given that `request_raw` returns an `awaitable` when invoked and so must do `do_request`, that we can just omit the `await` and the `async`. Interesting, good point. In any case, I think is more clear with the `async` and `await` so that it is explicit that everything is async.

[the remark this answers is not in the document]
Comment (ejulio) [reply scrapy_autoextract/providers.py:116]: Yeah, that was my idea here.
Since it already returns the coroutine, we don't need to make it async and await.
Though, thinking now, it makes sense to make it explicit and easier to understand the API..

File: README.rst [hunk 10]
```diff
@@ -69,13 +71,13 @@ The providers
 Another way of consuming AutoExtract API is using the Page Objects pattern
 proposed by the `web-poet`_ library and implemented by `scrapy-poet`_.
 
-Page Objects their returned Items are defined by the `autoextract-poet`_
+Page Objects returned Items are defined by the `autoextract-poet`_
```
Comment (kmike) [inline README.rst:74]:
```suggestion
Items returned by Page Objects are defined in the `autoextract-poet`_
```

File: scrapy_autoextract/task_manager.py [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+        async def my_task(sleep):
+          await asyncio.sleep(sleep)
+          return "Finished"
+
+        manager = TaskManager()
+        for i in range(10:
+          print(await manager.run(my_task(i)))
+
+        await asyncio.sleep(5)
+        manager.cancel_all()
```
Comment (kmike) [inline scrapy_autoextract/task_manager.py:40]:
```suggestion
        async def my_task(sleep):
            await asyncio.sleep(sleep)
            return "Finished"

        manager = TaskManager()
        for i in range(10):
            print(await manager.run(my_task(i)))

        await asyncio.sleep(5)
        manager.cancel_all()
```

File: scrapy_autoextract/providers.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (kmike) [reply scrapy_autoextract/providers.py:54]: `autoextract_account_disabled` does sounds like a better name for scrapy-autoextract. I'm fine with merging without making this changee, but I think we should change it eventually.

File: scrapy_autoextract/providers.py [hunk 12]
```diff
@@ -1,114 +1,229 @@
...
+        self.aiohttp_session = None
+        self.crawler.signals.connect(self.close_aiohttp_session,
+                                     signal=signals.spider_closed)
+        retries_count = self.settings.getint(
+                "AUTOEXTRACT_MAX_QUERY_ERROR_RETRIES", 3)
+        self.common_request_kwargs = dict(
+            api_key=self.settings.get("AUTOEXTRACT_USER"),
+            endpoint=self.settings.get("AUTOEXTRACT_URL"),
+            max_query_error_retries=retries_count,
+            session=self.aiohttp_session
```
Comment (kmike) [inline scrapy_autoextract/providers.py:93]: I don't think it works as intended: session is set to None here, and when self.aiohttp_session is created later, self.common_request_kwargs still has session=None. It seems `common_request_kwargs` may be better implemented as a property.

File: scrapy_autoextract/providers.py [hunk 13]
```diff
@@ -1,114 +1,229 @@
...
+                # html is requested only a single time to save resources
+                should_request_html = is_html_required and is_first_request
+                slot = self.get_per_domain_concurrency_slot(request)
+                ae_request = self.get_filled_request(
+                    request,
+                    provided_cls,
+                    should_request_html
+                )
+                if not ae_request:
+                    raise ValueError("Unexpected empty return from method get_filled_request")
```
Comment (kmike) [inline scrapy_autoextract/providers.py:192]: Why is this needed, won't it fail at some other place if get_filled_request is buggy?

File: scrapy_autoextract/providers.py [hunk 14]
```diff
@@ -1,114 +1,229 @@
...
+                awaitable = self.do_request(**{
+                    'query': [ae_request],
+                    'agg_stats': request_stats,
+                    **self.common_request_kwargs,
+                })
```
Comment (kmike) [inline scrapy_autoextract/providers.py:197]: Intresting - why isn't this working?

```python
awaitable = self.do_request(
    query=[ae_request],
    agg_stats=request_stats,
    **self.common_request_kwargs,
)

```

**Comment by kmike** [comment]
Looks good - great job @ivanprado! 
I've left a few comments & suggestions, but +1 to merge after fixing an issue with asyncio sessions not being used (//cc @sortafreel).

File: scrapy_autoextract/providers.py (hunk 5 above)
Comment (kmike) [reply scrapy_autoextract/providers.py:133]: If the behavior is so different, I wonder if it'd be better not to reuse CONCURRENT_REQUESTS_PER_DOMAIN option, but add a different option, AUTOEXTRACT_CONCURRENT_REQUESTS_DOMAIN.

Comment (kmike) [reply scrapy_autoextract/providers.py:133]: Can be done later though.

Comment (sortafreel) [reply scrapy_autoextract/providers.py:133]: Implement ``AUTOEXTRACT_CONCURRENT_REQUESTS_PER_DOMAIN`` setting to control the concurrency of AutoExtract requests separately for a more flexible configuration. If not set, the provider will search for the `CONCURRENT_REQUESTS_PER_DOMAIN` setting instead. Defaults to ``8``.

**Review by sortafreel — APPROVED** [review]

File: scrapy_autoextract/providers.py [hunk 15]
```diff
@@ -55,11 +55,17 @@ def _stop_if_account_disabled(exception: Exception, crawler: Crawler):
 
 
 def get_concurrent_requests_per_domain(settings: Settings):
-    """Return the configured concurrent request per domain from settings"""
-    concurrency = settings.getint("CONCURRENT_REQUESTS_PER_DOMAIN", 8)
+    """Return the configured AutoExtract concurrent request per domain from settings"""
+    limit_name = "AUTOEXTRACT_CONCURRENT_REQUESTS_PER_DOMAIN"
+    concurrency = settings.getint(limit_name, -1)
+    # If no AutoExtract-specific limit is provided - use the default one
+    if concurrency is -1:
```
Comment (kmike) [inline scrapy_autoextract/providers.py:62]:
```suggestion
    if concurrency == -1:
```

It is a CPython implementation detail that some int numbers are always the same object; better not to rely on it.

File: scrapy_autoextract/providers.py [hunk 16]
```diff
@@ -55,11 +55,17 @@ def _stop_if_account_disabled(exception: Exception, crawler: Crawler):
...
+        limit_name = "CONCURRENT_REQUESTS_PER_DOMAIN"
+        concurrency = settings.getint(limit_name, 8)
     if concurrency < 1:
-        raise ValueError("Invalid 'CONCURRENT_REQUESTS_PER_DOMAIN' "
+        raise ValueError(f"Invalid '{limit_name}' "
                          f"value: {concurrency}")
+    logger.info(f"Setting `concurrent requests per domain` limit to {concurrency}.")
```
Comment (kmike) [inline scrapy_autoextract/providers.py:68]: It'd be better not to log this line here - this function gets the concurrent requests to be used, not sets it. I'm also not sure it should be an INFO log message - what do you think about changing it to DEBUG? 

Option value is logged anyways.

File: README.rst [hunk 17]
```diff
@@ -207,9 +208,11 @@ Middleware settings
 Provider settings
 -----------------
 
-- ``AUTOEXTRACT_USER`` [optional] is your AutoExtract API key. Defaults to ``SCRAPINGHUB_AUTOEXTRACT_KEY`` environment variable.
+- ``AUTOEXTRACT_USER`` [optional] is your AutoExtract API key. If not set, it is
+  taken from SCRAPINGHUB_AUTOEXTRACT_KEY environment variable.
 - ``AUTOEXTRACT_URL`` [optional] the AutoExtract service url. Defaults to the official AutoExtract endpoint.
 - ``AUTOEXTRACT_MAX_QUERY_ERROR_RETRIES`` [optional] Max number of retries for Query-level errors. Defaults to ``3``.
+- ``AUTOEXTRACT_CONCURRENT_REQUESTS_PER_DOMAIN`` [optional] Max number of concurrent requests per domain. If not set, the provider will search for the `CONCURRENT_REQUESTS_PER_DOMAIN` setting instead. Defaults to ``8``.
```
Comment (kmike) [inline README.rst:215]: Nitpick: this setting defaults to ``CONCURRENT_REQUESTS_PER_DOMAIN``, not to 8. 
8 is a default value for ``CONCURRENT_REQUESTS_PER_DOMAIN``.

File: scrapy_autoextract/providers.py (hunk 5 above)
Comment (ivanprado) [reply scrapy_autoextract/providers.py:143]: Thank you @sortafreel for this improvement. However, the new behaviour is that if both `AUTOEXTRACT_CONCURRENT_REQUESTS_PER_DOMAIN` and `CONCURRENT_REQUESTS_PER_DOMAIN` are missing, an error is raised. But it should just default to 8 instead. I see that this was lost in this commit: https://github.com/scrapinghub/scrapy-autoextract/pull/16/commits/692fbe430ea85c478ba50fd9f35be1956610fca8

**Comment by ivanprado** [comment]
Thank you @sortafreel for taking care of that!

File: scrapy_autoextract/providers.py (hunk 5 above)
Comment (kmike) [reply scrapy_autoextract/providers.py:143]: If I'm not mistaken, `CONCURRENT_REQUESTS_PER_DOMAIN` can't be missing, as it is a built-in Scrapy setting; default is set here: https://github.com/scrapy/scrapy/blob/0a1e2fefab04984a0f3c2b470346fd9f2ffde65b/scrapy/settings/default_settings.py#L44. So I adviced @sortafreel to remove the default 8, to avoid duplication.

Comment (ivanprado) [reply scrapy_autoextract/providers.py:143]: Ouch. I see, thank you @kmike . So I was wrong. I think I saw some code in the Scrapy like `self.settings.getint('CONCURRENT_REQUESTS_PER_DOMAIN', 8)` but I cannot find it now so probably I'm confusing things.
